### PR TITLE
Implement Performance tests for System.Threading.

### PR DESCRIPTION
**Pull request by pjbgf**
Improves Performance test coverage from #18848.

### CODE REVIEW

**Comment by danmosemsft** [comment]
None of these interlocked tests involve a second thread. Is it useful/practical to introduce another thread? @stephentoub ?

**Comment by pjbgf** [comment]
@danmosemsft The Interlocked class is already tested in a multi-threaded context on InterlockedTests.cs - from a functional point of view. 
For the micro benchmarks I thought that the overhead of adding threads would impact the results, as the execution would no longer be completely isolated.

**Comment by danmosemsft** [comment]
I think you're right. I am not sure how likely these tests would be to see a regression, since it seems they basically just call through to the OS. It would be interesting to see how different OS performance varied though.

**Comment by stephentoub** [comment]
Are these tests stable and measuring the right things? i.e. do they do enough work per iteration to have repeatable measurements on each run, and are they actually measuring the cost of the functions under test?  e.g. In code like:
```C#
using (iteration.StartMeasurement())
{
    Volatile.Read(ref location);
}
````
I'd expect the cost of the Volatile.Read to be dwarfed by the costs of the using around it, at least on x86/x64.

**Comment by kouvel** [comment]
It might be good to use an InnerIterationCount as described here: https://github.com/Microsoft/xunit-performance

File: src/System.Threading/tests/Performance/Perf.Interlocked.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        [Benchmark]
+        public static void IncrementDecrement_int()
+        {
+            int integer32 = 0;
+            foreach (var iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                {
+                    Interlocked.Increment(ref integer32);
+                    Interlocked.Decrement(ref integer32);
```
Comment (kouvel) [inline src/System.Threading/tests/Performance/Perf.Interlocked.cs]: Would be good to do a fixed number of same operations per iteration such that the result is comparable across benchmarks, overflow shouldn't be an issue

**Comment by pjbgf** [comment]
@stephentoub thank you for the feedback. I refactored the tests and now they are stable and measuring the right things.

**Comment by kouvel** [comment]
I wonder if `Benchmark.Iterate()` would get inlined. I doesn't seem to be decorated with attributes to encourage inlining. If it's not inlined, I guess it would have to make a call for each inner iteration, which would also come into the measurement. I wonder if using the loops directly would produce better numbers. Maybe you can try on one simple interlocked or volatile and see if it is slowing down the test significantly?

**Comment by pjbgf** [comment]
@kouvel I does not look like it's being inlined. I have just tested both approaches and the variance is minimum, sometimes going up others down.

**Comment by kouvel** [comment]
Ok sounds good, LGTM

**Comment by kouvel** [comment]
@dotnet-bot test Innerloop OSX10.12 Debug x64 Build and Test

File: src/System.Threading/tests/Performance/Perf.Volatile.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    public class Perf_Volatile
+    {
+        [Benchmark(InnerIterationCount = 2000)]
+        public void Read_double()
+        {
+            double location = 0;
+
+            Benchmark.Iterate(() =>
+            {
+                Volatile.Read(ref location);
```
Comment (stephentoub) [inline src/System.Threading/tests/Performance/Perf.Volatile.cs]: I'm still suspect of this.  I'd expect the delegate invocation to significantly outweigh the cost of the volatile read.  Is there any measurable difference whether or not the Read is there or removed?

Comment (pjbgf) [inline src/System.Threading/tests/Performance/Perf.Volatile.cs]: @stephentoub You are absolutely right, I have just ran some tests and removing the delegate approach it gets extremely cheaper. I will refactor accordingly, thanks again.

**Comment by pjbgf** [comment]
Removed all the Benchmark.Iterate and also added performance tests for SpinLock.

**Comment by stephentoub** [comment]
@pjbgf, did you mean to close this?

**Comment by pjbgf** [comment]
@stephentoub sorry it is my first contribution here. What's the next steps to get it merged?

**Comment by kouvel** [comment]
I think we can merge after the tests complete. Does anyone else want to take a look?

**Comment by karelz** [comment]
@kouvel one MS code review is sufficient. Unless you want someone's specific attention ... if you feel confident, your word is enough.

@pjbgf I think you will need to rebase as there are conflicts: This is what I see "*This branch cannot be rebased due to conflicts*"

**Comment by kouvel** [comment]
@karelz I see, I was told the contrary before so I didn't realize that, thanks for clarifying

**Pull request merged**